### PR TITLE
Intro23

### PR DESCRIPTION
**Pull request by DesyncTheThird**

### CODE REVIEW

File: intro23.lean [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  . linarith
+  . exact h.right
+  done
+
+-- n; {4,5,6,...,n}
+-- c; cardinality of subsets
+def all_c_subsets_satisfy_tri_prop (c n : ℕ) : Prop :=
+  ∀(X:Finset ℕ), (X ⊆ four_to_n n) ∧ (card X = c) → tri_prop X
+
+theorem intro23_254counterexample :
```
Comment (mgsium) [inline intro23.lean:39]: A version using `simp`:
```lean
theorem intro23_254counterexample :
  ¬all_c_subsets_satisfy_tri_prop 10 254 := by
  let Y : Finset ℕ := {4, 5, 9, 14, 23, 37, 60, 97, 157, 254}
  simp [all_c_subsets_satisfy_tri_prop]
  use Y
  repeat (any_goals constructor) <;> simp
  done
```

File: intro23.lean [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  (Finset.range (b)) \ (Finset.range (a))
+
+def four_to_n (n : ℕ) : Finset ℕ :=
+  range_from_to 4 (n+1)
+
+lemma four_to_n_subset {n : ℕ} : four_to_n n ⊆ four_to_n (n+1) := by
+  unfold four_to_n range_from_to
+  intro x h
+  simp at *
+  constructor
```
Comment (mgsium) [inline intro23.lean:29]: `linarith` seems to close both goals here : `constructor <;> linarith`

File: intro23.lean [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  use Y
+  repeat (any_goals constructor)
+  . apply Finset.Subset.trans h.1.1
+    exact four_to_n_subset
+  . exact h.1.2
+  . exact h.2
+  done
+
+theorem intro23 (n : ℕ)
+  : n≥254 → n∉{i:ℕ | all_c_subsets_satisfy_tri_prop 10 i} := by
```
Comment (mgsium) [inline intro23.lean:67]: if we write the goals as `254 ≤ n → ¬ all_c_subsets_satisfy_tri_prop 10 n` we can avoid the `simp`

File: intro23.lean [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  simp
+  intro g
+  induction' n with d hd
+  . contradiction
+  . by_cases 254 = succ d
+    . rw [← h]
+      exact intro23_254counterexample
+    . have : d ≥ 254 := by
```
Comment (mgsium) [inline intro23.lean:75]: this can be written inline as `have : d ≥ 254 := lt_succ.mp $ Nat.lt_of_le_of_ne g h`

File: intro23.lean [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import Mathlib.Tactic
+
+open List
+open Set
+open Finset
+open Nat
+
+set_option maxHeartbeats 200000
```
Comment (mgsium) [inline intro23.lean:8]: do we need this?

File: intro23.lean (hunk 1 above)
Comment (DesyncTheThird) [reply intro23.lean:35]: Left unchanged for readability

**Review by dodoturkoz — APPROVED** [review]

**Review by mgsium — APPROVED** [review]

**Pull request merged**